File: grpc-spring/src/main/kotlin/tech/takenoko/grpcspring/repository/PlayerRepository.kt
package tech.takenoko.grpcspring.repository

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.withContext
import org.springframework.stereotype.Repository
import tech.takenoko.grpcspring.model.Vector3
import java.util.*


@Repository
class PlayerRepository {
    companion object {
        private val state = mutableMapOf<String, PlayerModel>()
        private val observers = mutableMapOf<String, MutableList<PlayerObserver>>()
    }

    suspend fun change(uuid: String, position: Vector3, rotation: Vector3) {
        state[uuid] = PlayerModel(position, rotation)
        withContext(Dispatchers.IO) {
            observers[uuid]?.forEach { it.update(state[uuid]) }
        }
    }

    suspend fun addObserver(uuid: String, o: PlayerObserver) {
        observers[uuid] = observers.getOrDefault(uuid, mutableListOf())
        observers[uuid]?.add(o)
    }

    suspend fun removeObserver(uuid: String, o: PlayerObserver) {
        observers[uuid]?.remove(o)
    }

    data class PlayerModel(val position: Vector3, val rotation: Vector3)

    interface PlayerObserver {
        suspend fun update(player: PlayerModel?)
    }
}

File: grpc-magic-onion-server/repository/PlayerRepository.cs
using System.Collections.Generic;
using UnityEngine;

namespace grpc_magic_onion_server.repository
{
    public class PlayerRepository
    {
        private static Dictionary<string, PlayerModel> state = new Dictionary<string, PlayerModel>();
        private static Dictionary<string, List<IPlayerObserver>> observers = new Dictionary<string, List<IPlayerObserver>>();
        
        public void Change(string uuid, Vector3 position, Vector3 rotation)
        {
            if (!state.ContainsKey(uuid)) state.Add(uuid, new PlayerModel());
            state[uuid] = new PlayerModel { position = position, rotation = rotation };
            // foreach (var playerObserver in observers[uuid]) playerObserver.update(state[uuid]);
        }

        public void AddObserver(string uuid, IPlayerObserver o)
        {
            if (observers.ContainsKey(uuid)) observers.Add(uuid, new List<IPlayerObserver>());
            observers[uuid]?.Add(o);
        }

        public void RemoveObserver(string uuid, IPlayerObserver o)
        {
            observers[uuid]?.Remove(o);
        }

        public class PlayerModel
        {
            public Vector3 position;
            public Vector3 rotation;
        }

        public interface IPlayerObserver
        {
            void update(PlayerModel player);
        }
    }
}